### PR TITLE
Improved logic of sharing industry production between stations

### PR DESCRIPTION
**Pull request by ldpl**
Since the dawn of time Industries in TTD only share cargo between two stations max. This behavior doesn't seem to be based on any logic and leads to some issues in mp as it can be easily abused. 

So this patch:
1. Removes the limit of 2 stations.
2. Shares cargo between companies first and then between stations of each company separately to prevent exploiting it by spaming station.

Shares are still based on station rating like it was before and, in fact, this patch doesn't change behaviour at all if there are less than 3 stations around (except for negligible rounding losses).

I didn't add a new setting for this change as IMO it only fixes exploits/limitations and doesn't change the gameplay logic. So I can't imagine why would anyone ever want to opt-out of it.

### CODE REVIEW

**Comment by ldpl** [comment]
Hm, apparently there was PR #7184 that got completely derailed and rejected. So, just to emphasize some points.

1. This PR doesn't change the way cargo is delivered from stations to industries which is indeed questionable.
2. This implementation has much better performance (though I haven't seen any evidence of this function being particularly time critical).
3. It doesn't try to invent a new way to determine share sizes and just replicates original logic of using station ratings even if it somewhat sucks.

It only solves one specific quirk of game mechanics that so far is only known to confuse people and/or being abused. For example, currently there is no such thing as competition between three companies on one industry.  Third station won't even get any cargo unless it uses another exploit to start acceptance.

File: src/station_cmd.cpp [hunk 1]
```diff
@@ -4014,68 +4014,68 @@ const StationList *StationFinder::GetStations()
 	return &this->stations;
 }
 
+static bool CanMoveGoodsToStation(const Station *st, CargoID type) {
+	if (st->owner != OWNER_NONE && st->town->exclusive_counter > 0 && st->town->exclusivity != st->owner) return false;
```
Comment (nielsmh) [inline src/station_cmd.cpp:4020]: Missing comment on this line.
Also, with this being its own function now I'd suggest moving the end-of-line comments to the line before each check, to improve readability.

File: src/station_cmd.cpp [hunk 2]
```diff
@@ -4014,68 +4014,68 @@ const StationList *StationFinder::GetStations()
...
+			if (r > best_rating) best_rating = r;
 		}
+		company_sum[st->owner] += r;
 	}
 
 	/* no stations around at all? */
-	if (st1 == nullptr) return 0;
+	if (best_rating == 0) return 0;
 
 	/* From now we'll calculate with fractal cargo amounts.
```
Comment (nielsmh) [inline src/station_cmd.cpp]: I think this should be "fractional", not "fractal".

File: src/station_cmd.cpp [hunk 3]
```diff
@@ -4014,68 +4014,68 @@ const StationList *StationFinder::GetStations()
...
-	 * of execution would be undefined and that could cause desyncs with callbacks. */
-	return moved + UpdateStationWaiting(st2, type, worst_cargo, source_type, source_id);
+		return UpdateStationWaiting(any_station, type, amount, source_type, source_id);
+	}
+
+	uint moving = 0;
+	for (Station *st : *all_stations) {
+		if (!CanMoveGoodsToStation(st, type)) continue;
+		/* Ignore rounding loss as it's only 1/256 of unit per station in the worst case */
+		uint station_amount = amount * company_best[st->owner] * st->goods[type].rating / best_sum / company_sum[st->owner];
```
Comment (nielsmh) [inline src/station_cmd.cpp]: It's not immediately obvious this expression won't allow more than the amount of cargo to distribute to be distributed. Can you write something explaining how the sum of all these fractions never exceeds 1?

**Comment by ldpl** [comment]
After a discussion in IRC I changed implementation so that it takes cargo bits that would've been lost previously to rounding issues and distributes them among the best rated stations, similarly to what original implementation did. 

Also I left CanMoveGoodsToStation as a separate function for better readability even though I don't need to call it twice anymore. 

Notably it works about as fast as original implementation in the most common cases of 0 or 1 stations. For the other cases it's about 6 times slower but that seems to be ok as they are quite rare when playing normally and it's probably going to waste even more time in StationFinder anyway. 

P.S. Can also be similarly optimized for 2 stations I just didn't see any particular reason to complicate it even more.

File: src/station_cmd.cpp [hunk 4]
```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
...
-	 * of rounding issues. */
-	uint worst_cargo = amount * best_rating2 / (best_rating1 + best_rating2);
-	assert(worst_cargo <= (amount - worst_cargo));
-
-	/* And then send the cargo to the stations! */
-	uint moved = UpdateStationWaiting(st1, type, amount - worst_cargo, source_type, source_id);
-	/* These two UpdateStationWaiting's can't be in the statement as then the order
-	 * of execution would be undefined and that could cause desyncs with callbacks. */
-	return moved + UpdateStationWaiting(st2, type, worst_cargo, source_type, source_id);
+uint MoveGoodsToStation(CargoID type, uint amount, SourceType source_type, SourceID source_id, const StationList *all_stations)
```
Comment (LordAro) [inline src/station_cmd.cpp:4038]: This whole function has managed to use spaces for indentation - this is why the commit checker is failing - pls fix

File: src/station_cmd.cpp [hunk 5]
```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
...
-		if (st->goods[type].rating == 0) continue; // Lowest possible rating, better not to give cargo anymore
-
-		if (_settings_game.order.selectgoods && !st->goods[type].HasVehicleEverTriedLoading()) continue; // Selectively servicing stations, and not this one
-
-		if (IsCargoInClass(type, CC_PASSENGERS)) {
-			if (st->facilities == FACIL_TRUCK_STOP) continue; // passengers are never served by just a truck stop
-		} else {
-			if (st->facilities == FACIL_BUS_STOP) continue; // non-passengers are never served by just a bus stop
-		}
+static bool CanMoveGoodsToStation(const Station *st, CargoID type) {
```
Comment (LordAro) [inline src/station_cmd.cpp]: opening `{` should be on its own line

File: src/station_cmd.cpp [hunk 6]
```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
...
-
-	/* And then send the cargo to the stations! */
-	uint moved = UpdateStationWaiting(st1, type, amount - worst_cargo, source_type, source_id);
-	/* These two UpdateStationWaiting's can't be in the statement as then the order
-	 * of execution would be undefined and that could cause desyncs with callbacks. */
-	return moved + UpdateStationWaiting(st2, type, worst_cargo, source_type, source_id);
+uint MoveGoodsToStation(CargoID type, uint amount, SourceType source_type, SourceID source_id, const StationList *all_stations)
+{
+    /* Return if nothing to do. Also the rounding below fails for 0. */
+    if (all_stations->size() == 0) return 0;
```
Comment (LordAro) [inline src/station_cmd.cpp]: could use `empty()` here, and in a few other places

File: src/station_cmd.cpp [hunk 7]
```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
...
+    if (amount == 0) return 0;
+
+    Station *first_station = nullptr;
+    typedef std::pair<Station *, uint> StationInfo;
+    std::vector<StationInfo> used_stations;
+
+    for (Station *st : *all_stations) {
+        if (!CanMoveGoodsToStation(st, type)) continue;
+
+        /* Avoid allocating a vector if there is only one station to significantly
```
Comment (LordAro) [inline src/station_cmd.cpp]: Can you provide some numbers on this optimisation? I feel like a `.reserve(1)` with the declaration of `used_stations` would have the same result, and have significantly simpler code

File: src/station_cmd.cpp [hunk 8]
```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
...
+        }
+        used_stations.emplace_back(std::make_pair(st, 0));
+    }
+
+    /* no stations around at all? */
+    if (first_station == nullptr) return 0;
+
+    if (used_stations.size() == 0) {
+        /* only one station around */
+        amount *= (first_station->goods[type].rating + 1);
```
Comment (LordAro) [inline src/station_cmd.cpp]: no need for parens here (and you don't use them on L4090 below)

**Review by LordAro — CHANGES_REQUESTED** [review]
Nothing too major

File: src/station_cmd.cpp (hunk 7 above)
Comment (ldpl) [reply src/station_cmd.cpp]: reserve(1) is about 3x slower for 1 station on gcc. Should be similar on other compilers as empty vectors are more likely to be optimized than vectors of size 1.

**Comment by ldpl** [comment]
FWIW here is the simple code I'm using to roughly check performance. 
[move-goods-to-station.zip](https://github.com/OpenTTD/OpenTTD/files/4048158/move-goods-to-station.zip)

**Review by LordAro — APPROVED** [review]

**Pull request merged**